libraries/botbuilder-applicationinsights/src/applicationInsightsTelemetryClient.ts

```diff
@@ -0,0 +1,144 @@
+// This allows AppInsights to automatically apply the appropriate context objects deep inside the async/await chain.
+const correlationContextManager = require('applicationinsights/out/AutoCollection/CorrelationContextManager').CorrelationContextManager;
+const origGetCurrentContext = correlationContextManager.getCurrentContext;
+
+function getCurrentContext(){
+  return ns.get('ctx') || origGetCurrentContext();
+}
+
+// Overwrite the built-in getCurrentContext() method with a new one.
+correlationContextManager.getCurrentContext = getCurrentContext;
+
+export const ApplicationInsightsWebserverMiddleware = function (req, res, next) {
+
+  // Check to see if the request contains an incoming request.
+  // If so, set it into the Application Insights context.
+  const activity = req.body;
+  if (activity && activity.id) {
+    const context = appInsights.getCorrelationContext();
+    context.activity = req.body;
+  }
+
+  ns.bindEmitter(req);
+  ns.bindEmitter(res);
+  ns.run(function(){
+    ns.set('ctx', origGetCurrentContext());
+    next();
+  });
+
+};
+
+
+/* ApplicationInsightsTelemetryClient Class
+ * This is a wrapper class around the Application Insights node client.
+ * This is primarily designed to be used alongside the WaterfallDialog telemetry collection.
+ * It provides a pre-configured App Insights client, and wrappers around
+ * the major tracking functions, allowing it to conform to Botbuilder's generic BotTelemetryClient interface.
+ * To use it, create pass in an instrumentation key:
+ * 
+ * ```
+ * const myDialog = new WaterfallDialog('my_dialog', steps);
+ * const appInsightsClient = new ApplicationInsightsTelemetryClient(my_instrumentation_key);
+ * myDialog.telemetryClient = appInsightsClient;
+ * ```
+ */
+export class ApplicationInsightsTelemetryClient implements BotTelemetryClient {
+
+    private _telemetryClient: any;
+    private _configuration: any;
+
+    /* The settings parameter is passed directly into appInsights.setup().
+     * https://www.npmjs.com/package/applicationinsights#basic-usage
+     * This function currently takes an app insights instrumentation key only.
+     */
+    constructor(settings: any) {
+
+        this._configuration = appInsights.setup(settings);
+
+        this._configuration.setAutoDependencyCorrelation(true)
+        .setAutoCollectRequests(true)
+        .setAutoCollectPerformance(true)
+        .setAutoCollectExceptions(true)
+        .setAutoCollectDependencies(true)
+        .start(); 
+
+        this._telemetryClient = appInsights.defaultClient;
+
+        this._telemetryClient.addTelemetryProcessor(addBotIdentifiers);
+    }
+
+    /* configuration() 
+     * Provides access to the Application Insights configuration that is running here.
+     * Allows developers to adjust the options, for example:
+     * `appInsightsClient.configuration.setAutoCollectDependencies(false)`
+     */
+    get configuration() {

```

Comment:
Intellisense will not work here since `_configuration` does not have a type. Is this what we want?